src/vs/workbench/browser/parts/panel/panelPart.ts

```diff
@@ -144,47 +153,162 @@ export class PanelPart extends CompositePart<Panel> implements IPanelService {
 
-		// Panel open/close
+		// Activate on panel open
 		this._register(this.onDidPanelOpen(({ panel }) => this.onPanelOpen(panel)));
+
+		// Deactivate on panel close
 		this._register(this.onDidPanelClose(this.onPanelClose, this));
 
-		// Panel register/deregister
-		this._register(this.registry.onDidRegister(panelDescriptor => this.compositeBar.addComposite(panelDescriptor)));
-		this._register(this.registry.onDidDeregister(panelDescriptor => {
-			this.compositeBar.hideComposite(panelDescriptor.id);
-			this.removeComposite(panelDescriptor.id);
+		// Extension registration
+		let disposables = this._register(new DisposableStore());
+		this._register(this.extensionService.onDidRegisterExtensions(() => {
+			disposables.clear();
+			this.onDidRegisterExtensions();
+			this.compositeBar.onDidChange(() => this.saveCachedPanels(), this, disposables);
+			this.storageService.onDidChangeStorage(e => this.onDidStorageChange(e), this, disposables);
 		}));
+	}
 
-		// Activate panel action on opening of a panel
-		this._register(this.onDidPanelOpen(({ panel }) => {
-			this.compositeBar.activateComposite(panel.getId());
-			this.layoutCompositeBar(); // Need to relayout composite bar since different panels have different action bar width
-		}));
+	private onDidRegisterExtensions(): void {
+		this.removeNotExistingComposites();
 
-		// Deactivate panel action on close
-		this._register(this.onDidPanelClose(panel => this.compositeBar.deactivateComposite(panel.getId())));
+		this.saveCachedPanels();
+	}
 
-		// State
-		this.lifecycleService.when(LifecyclePhase.Eventually).then(() => {
-			this._register(this.compositeBar.onDidChange(() => this.saveCachedPanels()));
-			this._register(this.storageService.onDidChangeStorage(e => this.onDidStorageChange(e)));
-		});
+	private removeNotExistingComposites(): void {
+		const panels = this.getPanels();
+		for (const { id } of this.getCachedPanels()) { // should this value match viewlet (load on ctor)
+			if (panels.every(panel => panel.id !== id)) {
+				this.hideComposite(id);
+			}
+		}
+	}
+
+	private hideComposite(compositeId: string): void {
+		this.compositeBar.hideComposite(compositeId);
+
+		const compositeActions = this.compositeActions.get(compositeId);
+		if (compositeActions) {
+			compositeActions.activityAction.dispose();
+			compositeActions.pinnedAction.dispose();
+			this.compositeActions.delete(compositeId);
+		}
 	}
 
 	private onPanelOpen(panel: IPanel): void {
 		this.activePanelContextKey.set(panel.getId());
+
+		// This panel supports view containers
+		if (panel instanceof PaneCompositePanel) {
+			const foundPanel = this.panelRegistry.getPanel(panel.getId());
+			if (foundPanel) {
+				this.compositeBar.addComposite(foundPanel);
+			}
+		}
+
+		// Activate composite when opened
+		this.compositeBar.activateComposite(panel.getId());
+
+		// This panel supports view containers
+		if (panel instanceof PaneCompositePanel) {
```

Comment:
Same as above. You can remove this check because you check for view container after